In [1]:
# # Install dependencies
# !pip install nltk sklearn PyStemmer tqdm

In [2]:
# NOTE: commented out for speed -- GZ-Eurlex preprocessing/visualization, not needed for the semantic-atoms experiment below. Uncomment to use.
# import json
# 
# import nltk
# nltk.download('stopwords')
# from nltk.corpus import stopwords
# STOPWORDS = set(stopwords.words('english'))
# 
# import re
# import nltk
# import pickle as pkl
# import numpy as np
# from tqdm import tqdm
# import scipy.sparse as sp
# from sklearn.feature_extraction.text import TfidfVectorizer
# from sklearn.preprocessing import normalize
# import Stemmer
# 
# def tokenizer_word(string):
#     string = string.lower()
#     tokens = [w for w in re.split("[\W,_]+", string) if ((w != "") and (w not in STOPWORDS))]
#     tokens = stemmer.stemWords(tokens)
#     return tokens
# 
# stemmer = Stemmer.Stemmer('english')

In [3]:
# NOTE: commented out for speed -- GZ-Eurlex preprocessing/visualization, not needed for the semantic-atoms experiment below. Uncomment to use.
# def read_sparse_mat(filename):
#     with open(filename) as f:
#         nr, nc = map(int, f.readline().split(' '))
#         data = []; inds = []; indptr = [0]
#         for line in tqdm(f, total=nr):
#             row = list(map(lambda x: x.split(':'), line.split()))
#             inds.append(np.array(list(map(lambda x: int(x[0]), row))))
#             data.append(np.array(list(map(lambda x: float(x[1]), row))))
#             indptr.append(indptr[-1]+len(row))
#         spmat = sp.csr_matrix((np.concatenate(data), np.concatenate(inds), indptr), (nr, nc))
#         return spmat

In [4]:
# NOTE: commented out for speed -- GZ-Eurlex preprocessing/visualization, not needed for the semantic-atoms experiment below. Uncomment to use.
# trn_X_Y = read_sparse_mat(f'GZ-Eurlex-4.3K/trn_X_Y.txt')
# 
# trnX = [x.strip() for x in open(f'GZ-Eurlex-4.3K/raw/trn_X.txt')]
# tstX = [x.strip() for x in open(f'GZ-Eurlex-4.3K/raw/tst_X.txt')]
# Y = [x.strip() for x in open(f'GZ-Eurlex-4.3K/raw/Y.txt')]
# parent_Y = [x.strip() for x in open(f'GZ-Eurlex-4.3K/raw/Y.parent.txt')]
# 
# nnz = trn_X_Y.getnnz(0)
# seen_labels = np.where(nnz > 0)[0]
# seen_Y = [Y[i] for i in seen_labels]

In [5]:
# NOTE: commented out for speed -- GZ-Eurlex preprocessing/visualization, not needed for the semantic-atoms experiment below. Uncomment to use.
# # For bigger datasets (amazon-1m, wikipedia-1m) use larger MAX_FEATURES (500000/1000000 used in zestxml paper)
# MAX_FEATURES=50000

In [6]:
# NOTE: commented out for speed -- GZ-Eurlex preprocessing/visualization, not needed for the semantic-atoms experiment below. Uncomment to use.
# unigram_vectorizer = TfidfVectorizer(lowercase=False, tokenizer=tokenizer_word, ngram_range=(1, 1), max_df=0.5, norm=None, max_features=MAX_FEATURES)
# bigram_vectorizer = TfidfVectorizer(lowercase=False, tokenizer=tokenizer_word, ngram_range=(2, 2), max_df=0.5, norm=None, max_features=MAX_FEATURES)
# 
# %time unigram_trn_X_Xf = unigram_vectorizer.fit_transform(trnX)
# %time bigram_trn_X_Xf = bigram_vectorizer.fit_transform(trnX)
# 
# %time unigram_tst_X_Xf = unigram_vectorizer.transform(tstX)
# %time bigram_tst_X_Xf = bigram_vectorizer.transform(tstX)
# 
# trn_X_Xf = normalize(sp.hstack([unigram_trn_X_Xf, bigram_trn_X_Xf]))
# tst_X_Xf = normalize(sp.hstack([unigram_tst_X_Xf, bigram_tst_X_Xf]))
# Xf = np.concatenate([unigram_vectorizer.get_feature_names_out(), bigram_vectorizer.get_feature_names_out()])

In [7]:
# NOTE: commented out for speed -- GZ-Eurlex preprocessing/visualization, not needed for the semantic-atoms experiment below. Uncomment to use.
# unigram_vectorizer = TfidfVectorizer(lowercase=False, tokenizer=tokenizer_word, ngram_range=(1, 1), max_df=0.5, norm=None, max_features=MAX_FEATURES)
# bigram_vectorizer = TfidfVectorizer(lowercase=False, tokenizer=tokenizer_word, ngram_range=(2, 2), max_df=0.5, norm=None, max_features=MAX_FEATURES)
# parent_vectorizer = TfidfVectorizer(lowercase=False, norm=None)
# 
# %time unigram_Y_Yf = unigram_vectorizer.fit_transform(Y)
# %time bigram_Y_Yf = bigram_vectorizer.fit_transform(Y)
# %time parent_Y_Yf = parent_vectorizer.fit_transform(parent_Y)
# lbl_Y_Yf = sp.csr_matrix((np.full(len(seen_Y), unigram_vectorizer.idf_.max()), 
#                           np.arange(len(seen_Y)), 
#                           np.append([0], np.cumsum(nnz > 0))), 
#                          (len(Y), len(seen_Y)))
# 
# Y_Yf = normalize(sp.hstack([unigram_Y_Yf, bigram_Y_Yf, lbl_Y_Yf, parent_Y_Yf]))
# Yf = np.concatenate([unigram_vectorizer.get_feature_names_out(), 
#                      bigram_vectorizer.get_feature_names_out(), 
#                      [f'__label__{i}__{Y[i][:50]}' for i in range(len(Y)) if nnz[i] > 0], 
#                      [f'__parent__{yf}' for yf in parent_vectorizer.get_feature_names_out()]])

## Visualize

In [8]:
# NOTE: commented out for speed -- GZ-Eurlex preprocessing/visualization, not needed for the semantic-atoms experiment below. Uncomment to use.
# def get_text(x, text, X_Xf, sep=' ', K=-1, attr='bold underline'):
#     if K == -1: K = X_Xf[x].nnz
#     sorted_inds = X_Xf[x].indices[np.argsort(-X_Xf[x].data)][:K]
#     return '%d : \n'%x + sep.join(['%s(%.2f, %d)'%(_c(text[i], attr=attr), X_Xf[x, i], i) for i in sorted_inds])

In [9]:
# NOTE: commented out for speed -- GZ-Eurlex preprocessing/visualization, not needed for the semantic-atoms experiment below. Uncomment to use.
# class bcolors:
#     purple = '\033[95m'
#     blue = '\033[94m'
#     green = '\033[92m'
#     warn = '\033[93m' # dark yellow
#     fail = '\033[91m' # dark red
#     white = '\033[37m'
#     yellow = '\033[33m'
#     red = '\033[31m'
#     
#     ENDC = '\033[0m'
#     bold = '\033[1m'
#     underline = '\033[4m'
#     reverse = '\033[7m'
#     
#     on_grey = '\033[40m'
#     on_yellow = '\033[43m'
#     on_red = '\033[41m'
#     on_blue = '\033[44m'
#     on_green = '\033[42m'
#     on_magenta = '\033[45m'
#     
# def _c(*args, attr='bold'):
#     string = ''.join([bcolors.__dict__[a] for a in attr.split()])
#     string += ' '.join([str(arg) for arg in args])+bcolors.ENDC
#     return string

In [10]:
# NOTE: commented out for speed -- GZ-Eurlex preprocessing/visualization, not needed for the semantic-atoms experiment below. Uncomment to use.
# orig_Xf = [x.strip() for x in open(f'GZ-Eurlex-4.3K/Xf.txt')]
# orig_trn_X_Xf = read_sparse_mat(f'GZ-Eurlex-4.3K/trn_X_Xf.txt')
# 
# orig_Yf = [x.strip() for x in open(f'GZ-Eurlex-4.3K/Yf.txt')]
# orig_Y_Yf = read_sparse_mat(f'GZ-Eurlex-4.3K/Y_Yf.txt')

In [11]:
# NOTE: commented out for speed -- GZ-Eurlex preprocessing/visualization, not needed for the semantic-atoms experiment below. Uncomment to use.
# x = np.random.randint(len(trnX))
# 
# print(get_text(x, Xf, trn_X_Xf))
# print(get_text(x, orig_Xf, orig_trn_X_Xf))

In [12]:
# NOTE: commented out for speed -- GZ-Eurlex preprocessing/visualization, not needed for the semantic-atoms experiment below. Uncomment to use.
# y = np.random.randint(len(Y))
# 
# print(get_text(y, Yf, Y_Yf))
# print(get_text(y, orig_Yf, orig_Y_Yf))

## Sparse semantic atoms vs semantic-ID codes (zero-shot Reuters)

Tests whether a **sparse model can be semantic**: ZestXML's bilinear machinery is
representation-agnostic — it needs a sparse basis *shared by the doc and label sides*,
not necessarily a lexical one. We compare, inside the unmodified ZestXML binary:

1. **lexical** — stemmed unigram+bigram tf-idf (standard ZestXML recipe)
2. **+ atoms** — lexical ⊕ flat k-sparse semantic atoms (corpus tf-idf → LSA →
   k-sparse autoencoder; 512 atoms, k=16, weight 0.3), as `#atom<j>` features on
   **both** Xf and Yf (no underscore in the token, so the direct Xf↔Yf matcher fires)
3. **+ RQ codes** — same LSA embeddings quantized semantic-ID style (residual
   k-means, 4×32) through the identical harness — the ablation that isolates the
   *shape* of the basis

Dataset: GZXML-format zero-shot split of Reuters-21578 via NLTK (7,105 train /
3,019 test docs, 25 of 90 labels unseen). See `semantic_atoms/README.md` for the
full writeup; scripts are dataset-agnostic and run unchanged on GZ-Eurlex-4.3K.


In [ ]:
# Colab bootstrap — run this first on Colab; harmless no-op if the repo is already the cwd
import os
if not os.path.exists('semantic_atoms'):
    !git clone -b claude/semantic-ids-zestxml-othz84 https://github.com/hanialshater/zestxml.git
    %cd zestxml
    !make > /dev/null
    !pip -q install numpy scipy scikit-learn nltk torch tqdm
    !python3 -c "import nltk; [nltk.download(p, quiet=True) for p in ['reuters','stopwords']]"


In [13]:
# One-time setup (uncomment on first run):
# !pip install numpy scipy scikit-learn nltk torch tqdm
# !python3 -c "import nltk; [nltk.download(p) for p in ['reuters','stopwords']]"
# !make   # build the ZestXML binary


In [14]:
# Build the zero-shot Reuters dataset in GZXML format (~1 min, skips nothing if re-run)
!python3 semantic_atoms/build_reuters.py


unseen labels: ['alum', 'barley', 'cocoa', 'coffee', 'copper', 'cpi', 'gas', 'gold', 'heat', 'hog', 'income', 'iron-steel', 'livestock', 'money-supply', 'palm-oil', 'pet-chem', 'platinum', 'rapeseed', 'silver', 'soy-meal', 'soybean', 'strategic-metal', 'sugar', 'sunseed', 'yen']
dropping 664 train docs with only unseen labels


wrote GZXML-Datasets/GZ-Reuters-90: trn (7105, 29171), tst (3019, 29171), Y (90, 29261), 25 unseen labels


In [15]:
# Learn the k-sparse semantic atoms and write the augmented dataset (~1 min, CPU)
!python3 semantic_atoms/build_atoms.py GZXML-Datasets/GZ-Reuters-90 \
    --n-atoms 512 --k 16 --atom-wt 0.3 --out GZXML-Datasets/GZ-Reuters-90-atoms


fitting shared tf-idf + LSA on train docs + label texts...


  LSA explained variance: 0.486
training k-sparse AE (512 atoms, k=16)...


  sae epoch 0: recon mse 0.7237


  sae epoch 10: recon mse 0.3380


  sae epoch 20: recon mse 0.2533


  sae epoch 30: recon mse 0.2035


  sae epoch 39: recon mse 0.1707


512 atoms in use; mean active/doc: 16.0, /label: 15.7
writing augmented dataset to GZXML-Datasets/GZ-Reuters-90-atoms


done


In [16]:
# Ablation: RQ-style hierarchical codes through the identical harness
!cd semantic_atoms && python3 build_rq_codes.py ../GZXML-Datasets/GZ-Reuters-90 \
    --out ../GZXML-Datasets/GZ-Reuters-90-rqcodes


residual k-means: 4 levels x 32 centroids


  level 0: residual norm 0.795


  level 1: residual norm 0.749


  level 2: residual norm 0.714


  level 3: residual norm 0.683


wrote ../GZXML-Datasets/GZ-Reuters-90-rqcodes


In [17]:
# Train + predict all three variants (~15 s each on CPU)
HP = '-bilinear_classifier_cost 5 -bs_count 60 -bs_direct_wt 0.8 -bs_alpha 0.02 -shortyK 30'
for ds in ['GZ-Reuters-90', 'GZ-Reuters-90-atoms', 'GZ-Reuters-90-rqcodes']:
    !./run.sh {ds} train   {HP} > /dev/null 2>&1
    !./run.sh {ds} predict {HP} > /dev/null 2>&1
    print('done:', ds)


done: GZ-Reuters-90


done: GZ-Reuters-90-atoms


done: GZ-Reuters-90-rqcodes


In [18]:
# Evaluate with seen/unseen label splits
import sys
sys.path.insert(0, 'semantic_atoms')
from evaluate import evaluate

for name in ['GZ-Reuters-90', 'GZ-Reuters-90-atoms', 'GZ-Reuters-90-rqcodes']:
    evaluate(name)



=== GZ-Reuters-90 ===


overall  (3019 docs): P@1=87.68  P@3=35.35  P@5=22.40
seen     (2794 docs): P@1=94.56  P@3=37.40  P@5=23.03
unseen   (374 docs): P@1=74.87  P@3=28.79  P@5=17.27  R@5=78.25  R@10=78.25

=== GZ-Reuters-90-atoms ===


overall  (3019 docs): P@1=87.38  P@3=35.63  P@5=22.62
seen     (2794 docs): P@1=94.20  P@3=37.44  P@5=23.01
unseen   (374 docs): P@1=78.07  P@3=33.16  P@5=19.95  R@5=88.79  R@10=89.59

=== GZ-Reuters-90-rqcodes ===


overall  (3019 docs): P@1=87.21  P@3=35.36  P@5=22.44
seen     (2794 docs): P@1=94.20  P@3=37.43  P@5=23.05
unseen   (374 docs): P@1=76.20  P@3=31.11  P@5=18.88  R@5=84.43  R@10=84.43


### Reference results

| unseen labels (374 docs) | P@1 | P@3 | R@5 | R@10 |
|---|---|---|---|---|
| lexical baseline | 74.87 | 28.79 | 78.25 | 78.25 |
| + RQ codes (semantic-ID shaped) | 76.20 | 31.11 | 84.43 | 84.43 |
| + sparse atoms | **78.07** | **33.16** | **88.79** | **89.59** |

Seen P@1 stays ~94.2–94.6 in all three. The lexical model's unseen recall
saturates (R@5 = R@10): labels it can't lexically match are never surfaced at any
depth — the atoms fix exactly that. RQ codes help but lose to flat overcomplete
atoms on every unseen metric: tree-path codes are coarse per level and
non-compositional, while k-sparse atoms decompose an unseen label into many
*trained*, shared components.
